File: fsharp/nucleotide-count/NucleotideCount.fs
module NucleotideCount

let nucleotideCounts (strand : string) : Option<Map<char, int>> =
    let nucleotides = [| 'A'; 'C'; 'G'; 'T' |]

    if strand |> Seq.exists (fun ch -> not <| Array.contains ch nucleotides) then
        None
    else
        nucleotides
        |> Array.map (fun ch -> ch, strand |> Seq.filter ((=) ch) |> Seq.length)
        |> Map.ofArray
        |> Some

File: clojure/nucleotide-count/src/nucleotide_count.clj
(ns nucleotide-count)

(def nucleotides {\A 0, \T 0, \C 0, \G 0})

(defn count-of-nucleotide-in-strand [nucleotide strand]
  {:pre [(nucleotides nucleotide)]}
  (count (filter #{nucleotide} strand)))

(defn nucleotide-counts [strand]
  (into nucleotides (frequencies strand)))
